### DISCUSSION TITLE
Retrieve log streaming

### DISCUSSION THREAD

**Discussion by brandon-yilmaz** [sassoftware/saspy#457] [Q&A] [answered]
Hello everyone!

Is there any way to retrieve the SAS log in a streaming-mode?
The SASsession.submit method returns the log but only after the execution in SAS has already finished. 

Thanks !!

**Comment by tomweber-sas** [comment, marked as answer]
Well, no, not really. The method call doesn't 'return' to you until it's finished. Which means that unless you had multiple threads you couldn't even be getting the log while the submit was still executing in the other thread. And that's not how this works. Also, depending upon the access method, I don't necessarily get the log, from SAS, until after the code is finished executing, since those are API calls to SAS and they don't provide the output until after the code finishes. STDIO is getting it 'live', but it can't 'return' streaming unless you had another thread to be reading it. For debugging, you can add print() for the log in the loop in submit (for STDIO) so it would just write it out to the log as it's getting output. HTTP can't get any results till after all the code runs, so that won't work there. IOM is kinda in between so you might be able to print if anything comes back before the code finishes.

**Reply by brandon-yilmaz** [reply]
First of all, thank you so much for your response Tom!!
My first approach was through multiple Threads, but still impossible to get a 'stream' log because -as you mentioned- the API just returns the output when the execution is already finished.  
Thinking about using the STDIO method over SSH and  Threads.

Thanks again!
